### PR TITLE
feat(engine): Add scatter interval for rate-limited task execution

### PR DESCRIPTION
**Pull request by daryllimyt**
## Summary
Adds support for staggered scatter task execution with configurable intervals, enabling rate-limiting and controlled parallel execution.

## Changes
- Added `interval` parameter to `core.transform.scatter` action
- Added `delay` field to `Task` model (excluded from equality checks)
- Modified scheduler to respect task delays before execution
- Task delays are inherited by downstream tasks in scatter regions

## Usage
```yaml
- ref: scatter_requests
  action: core.transform.scatter
  args:
    collection: ${{ [1, 2, 3, 4, 5] }}
    interval: 0.5  # 0.5s delay between each task
```

With `interval: 0.5`, tasks are scheduled with delays:
- Task 0: 0s delay
- Task 1: 0.5s delay
- Task 2: 1.0s delay
- Task 3: 1.5s delay
- Task 4: 2.0s delay

### interval=3 
Produces the following Temporal task execution
image

Implementation doesn't propagate the `Task.delay` (expected behavior) when there are multiple actions in the stream.

https://github.com/user-attachments/assets/f714ec4d-1492-4ab9-9687-7e909649d952

## Tests
- 7 integration tests covering basic, nested, and edge cases
- 16 unit tests for model changes
- All tests passing ✅

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> Adds `interval` to `core.transform.scatter` to stagger tasks, introduces `Task.delay`, and updates scheduler to respect delays with extensive tests.
> 
> - **DSL/Scheduler**:
>   - `tracecat/dsl/models.py`: Add `Task.delay` (excluded from equality) and `ScatterArgs.interval`.
>   - `tracecat/dsl/scheduler.py`: Respect per-task delays (sleep before execute), inherit and normalize delays across queued tasks, and compute scatter delays as `index * interval`.
> - **Registry Interface**:
>   - `packages/tracecat-registry/tracecat_registry/core/transform.py`: Extend `core.transform.scatter` signature with optional `interval` arg.
> - **Tests**:
>   - Add integration tests for staggered scatter execution, nested scatters, zero/none interval, empty collections, and rate-limiting behavior (`tests/unit/test_scatter_interval.py`).
>   - Add unit tests for `Task.delay` and `ScatterArgs.interval` models (`tests/unit/test_scatter_interval_models.py`).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a1a3d23e2b0eee9456100e6676e8d7bfc5d18d0b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: tracecat/dsl/scheduler.py [hunk 1]
```diff
@@ -404,6 +405,13 @@ async def _schedule_task(self, task: Task) -> None:
 
             # 4) If we made it here, the task is reachable and not force-skipped.
 
+            # Respect the task delay if it exists. We need this to stagger tasks for scatter.
+            if task.delay > 0:
+                self.logger.info(
+                    "Task has delay, sleeping", task=task, delay=task.delay
+                )
+                await asyncio.sleep(task.delay)
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/scheduler.py]: <html><head></head><body><div class="prose text-pretty dark:prose-invert inline leading-relaxed break-words min-w-0 [word-break:break-word] prose-strong:font-medium"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">In Temporal, workflow.sleep is effectively the same as asyncio.sleep in that using asyncio.sleep in a Temporal workflow creates a durable, Temporal-managed timer rather than a simple local sleep. Temporal integrates with asyncio by providing a deterministic, durable asyncio event loop for workflows. When a workflow executes await asyncio.sleep(duration), it starts a durable timer managed by Temporal's server, which does not block the thread or worker and is resilient against crashes. The workflow can be resumed exactly where it left off when the timer completes, enabling durability and replay.</p>
<p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">There is no separate workflow.sleep function in the Temporal Python SDK; workflows just use asyncio.sleep to create these durable timers. This mechanism allows workflows to "sleep" for long periods without resource tying or blocking threads. So while asyncio.sleep is commonly a non-blocking delay in normal Python code, in the context of Temporal workflows it is a durable timer backed by the Temporal infrastructure.<span class="citation inline" rel="nofollow noopener" data-state="closed"><a rel="noopener" class="inline" target="_blank" href="https://temporal.io/blog/durable-distributed-asyncio-event-loop"><span class="relative -mt-px select-none whitespace-nowrap -top-px font-sans text-base text-foreground selection:bg-super/50 selection:text-foreground dark:selection:bg-super/10 dark:selection:text-super"><span class="text-3xs rounded-badge group min-w-4 cursor-pointer text-center align-middle font-mono tabular-nums py-[0.1875rem] leading-snug px-[0.3rem] [@media(hover:hover)]:hover:bg-super dark:[@media(hover:hover)]:hover:text-inverse [@media(hover:hover)]:hover:text-white border-subtlest ring-subtlest divide-subtlest bg-subtle"><span class="relative -mt-px inline-block align-middle max-w-[25ch] overflow-hidden">temporal</span><span class="ml-xs -mt-px mr-px inline-block align-middle"><span class="opacity-50">+3</span></span></span></span></a></span></p>
<p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">To summarize:</p>
<div class="group relative"><div class="w-full overflow-x-auto md:max-w-[90vw] border-subtlest ring-subtlest divide-subtlest bg-transparent">
Aspect | asyncio.sleep (normal Python) | workflow.sleep / asyncio.sleep (Temporal workflow)
-- | -- | --
Implementation | Non-blocking async delay using OS event loop | Durable Temporal timer integrated into workflow event handling
Thread blocking | No | No
Durability | No (process crash loses state) | Yes (workflow state is replayed and resumed after crash)
Resource usage | Uses event loop resources | Does not consume worker resources while sleeping
Function availability | asyncio.sleep | Use asyncio.sleep; no separate workflow.sleep in Temporal Python

</div><div class="bg-base border-subtler shadow-subtle pointer-coarse:opacity-100 right-xs absolute bottom-0 flex rounded-lg border opacity-0 transition-opacity group-hover:opacity-100 [&amp;&gt;*:not(:first-child)]:border-subtle [&amp;&gt;*:not(:first-child)]:border-l"><div class="flex" style="opacity: 1;"><button aria-label="Copy table" type="button" class="focus-visible:bg-subtle text-quiet hover:text-foreground aspect-square font-sans focus:outline-none outline-none outline-transparent transition duration-300 ease-out select-none items-center relative group/button font-semimedium justify-center text-center items-center rounded-lg cursor-pointer active:scale-[0.97] active:duration-150 active:ease-outExpo origin-center whitespace-nowrap inline-flex text-sm h-8 aspect-[9/8]" data-state="closed"><div class="flex items-center min-w-0 gap-two justify-center"><div class="flex shrink-0 items-center justify-center size-4"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="tabler-icon tabler-icon-copy tabler-icon"><path d="M7 7m0 2.667a2.667 2.667 0 0 1 2.667 -2.667h8.666a2.667 2.667 0 0 1 2.667 2.667v8.666a2.667 2.667 0 0 1 -2.667 2.667h-8.666a2.667 2.667 0 0 1 -2.667 -2.667z"></path><path d="M4.012 16.737a2.005 2.005 0 0 1 -1.012 -1.737v-10c0 -1.1 .9 -2 2 -2h10c.75 0 1.158 .385 1.5 1"></path></svg></div></div></button></div><div class="flex" style="opacity: 1;"><button aria-label="Download CSV" type="button" class="focus-visible:bg-subtle text-quiet hover:text-foreground aspect-square font-sans focus:outline-none outline-none outline-transparent transition duration-300 ease-out select-none items-center relative group/button font-semimedium justify-center text-center items-center rounded-lg cursor-pointer active:scale-[0.97] active:duration-150 active:ease-outExpo origin-center whitespace-nowrap inline-flex text-sm h-8 aspect-[9/8]" data-state="closed"><div class="flex items-center min-w-0 gap-two justify-center"><div class="flex shrink-0 items-center justify-center size-4"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="tabler-icon tabler-icon-download tabler-icon"><path d="M4 17v2a2 2 0 0 0 2 2h12a2 2 0 0 0 2 -2v-2"></path><path d="M7 11l5 5l5 -5"></path><path d="M12 4l0 12"></path></svg></div></div></button></div></div></div>
<p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">In practice, for Temporal Python SDK workflows, just use await asyncio.sleep to achieve the durable sleep behavior that is consistent with workflow execution semantics.<span class="citation inline" rel="nofollow noopener" data-state="closed"><a rel="noopener" class="inline" target="_blank" href="https://github.com/temporalio/sdk-python"><span class="relative -mt-px select-none whitespace-nowrap -top-px font-sans text-base text-foreground selection:bg-super/50 selection:text-foreground dark:selection:bg-super/10 dark:selection:text-super"><span class="text-3xs rounded-badge group min-w-4 cursor-pointer text-center align-middle font-mono tabular-nums py-[0.1875rem] leading-snug px-[0.3rem] [@media(hover:hover)]:hover:bg-super dark:[@media(hover:hover)]:hover:text-inverse [@media(hover:hover)]:hover:text-white border-subtlest ring-subtlest divide-subtlest bg-subtle"><span class="relative -mt-px inline-block align-middle max-w-[25ch] overflow-hidden">github</span><span class="ml-xs -mt-px mr-px inline-block align-middle"><span class="opacity-50">+3</span></span></span></span></a></span></p></div>

<ol>
  <li><a href="https://temporal.io/blog/durable-distributed-asyncio-event-loop">https://temporal.io/blog/durable-distributed-asyncio-event-loop</a></li>
  <li><a href="https://community.temporal.io/t/are-each-workflow-sleep-and-worflow-await-backed-by-a-thread-if-so-when-will-the-thread-be-relinquished/1526">https://community.temporal.io/t/are-each-workflow-sleep-and-worflow-await-backed-by-a-thread-if-so-when-will-the-thread-be-relinquished/1526</a></li>
  <li><a href="https://github.com/temporalio/sdk-python">https://github.com/temporalio/sdk-python</a></li>
  <li><a href="https://community.temporal.io/t/if-a-workflow-sleep-does-it-also-sleep-spawned-activities/6221">https://community.temporal.io/t/if-a-workflow-sleep-does-it-also-sleep-spawned-activities/6221</a></li>
  <li><a href="https://community.temporal.io/t/how-to-use-sleep-in-python-sdk/8773">https://community.temporal.io/t/how-to-use-sleep-in-python-sdk/8773</a></li>
  <li><a href="https://community.temporal.io/t/workflow-sleep-details/1161">https://community.temporal.io/t/workflow-sleep-details/1161</a></li>
  <li><a href="https://docs.temporal.io/develop/python/timers">https://docs.temporal.io/develop/python/timers</a></li>
  <li><a href="https://javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/workflow/Workflow.html">https://javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/workflow/Workflow.html</a></li>
  <li><a href="https://community.temporal.io/t/canceling-updating-a-temporal-sleep/12506">https://community.temporal.io/t/canceling-updating-a-temporal-sleep/12506</a></li>
  <li><a href="https://stackoverflow.com/questions/56729764/asyncio-sleep-vs-time-sleep">https://stackoverflow.com/questions/56729764/asyncio-sleep-vs-time-sleep</a></li>
</ol></body></html>

File: tests/unit/test_scatter_interval.py [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+        )
+
+    end_time = time.time()
+    elapsed = end_time - start_time
+
+    # Verify results
+    assert result["ACTIONS"]["gather"]["result"] == [2, 4, 6]
+
+    # With zero interval, should execute quickly
+    assert elapsed < 2.0
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tests/unit/test_scatter_interval.py:232]: deal with it later

File: tests/unit/test_scatter_interval.py [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+        )
+
+    end_time = time.time()
+    elapsed = end_time - start_time
+
+    # Verify results
+    assert result["ACTIONS"]["gather"]["result"] == [2, 3, 4, 5]
+
+    # Without interval, execution should be relatively fast (< 2 seconds for simple ops)
+    assert elapsed < 2.0, (
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tests/unit/test_scatter_interval.py:167]: deal with it later

File: tracecat/dsl/models.py [hunk 4]
```diff
@@ -365,12 +365,17 @@ class Task:
     """The task action reference"""
     stream_id: StreamID
     """The stream ID of the task"""
+    delay: float = field(default=0.0, compare=False)
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/models.py:368]: No. The scheduling path that handles “confluence” nodes already propagates whatever delay is present on the predecessor task. When _queue_tasks fans out from a task into its dependents, it explicitly constructs each downstream Task with delay=task.delay, preserving the incoming interval (see tracecat/dsl/scheduler.py:352-369). The only moment the delay is cleared is inside _schedule_task, right after the throttle has been honoured (asyncio.sleep(delay)), where the runner intentionally does task = replace(task, delay=0.0) so that the next hop is not delayed again (tracecat/dsl/scheduler.py:409-417). That behaviour existed before the comment and applies to all successors—including joins—so there isn’t an overlooked “multi-dependency” branch emitting fresh Task objects with the default delay.

The new delay field being marked compare=False simply keeps dictionary/set lookups (e.g. self.indegrees) stable; it doesn’t cause the scheduler to drop an inherited delay. The only places where we instantiate a Task without supplying delay are bookkeeping cases such as skip-stream placeholders or the parent gather wrapper (tracecat/dsl/scheduler.py:629, 647, 894, 1068), and those are unrelated to the rate-limiting that the scatter interval provides.

File: tracecat/dsl/scheduler.py [hunk 5]
```diff
@@ -404,6 +405,15 @@ async def _schedule_task(self, task: Task) -> None:
...
+                # Reset the delay to 0.0 so it doesn't propagate to the next task
+                task = replace(task, delay=0.0)
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/scheduler.py]: <html><head></head><body><div class="prose text-pretty dark:prose-invert inline leading-relaxed break-words min-w-0 [word-break:break-word] prose-strong:font-medium"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">In Temporal, workflow.sleep is effectively the same as asyncio.sleep in that using asyncio.sleep in a Temporal workflow creates a durable, Temporal-managed timer rather than a simple local sleep. Temporal integrates with asyncio by providing a deterministic, durable asyncio event loop for workflows. When a workflow executes await asyncio.sleep(duration), it starts a durable timer managed by Temporal's server, which does not block the thread or worker and is resilient against crashes. The workflow can be resumed exactly where it left off when the timer completes, enabling durability and replay.</p>
<p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">There is no separate workflow.sleep function in the Temporal Python SDK; workflows just use asyncio.sleep to create these durable timers. This mechanism allows workflows to "sleep" for long periods without resource tying or blocking threads. So while asyncio.sleep is commonly a non-blocking delay in normal Python code, in the context of Temporal workflows it is a durable timer backed by the Temporal infrastructure.<span class="citation inline" rel="nofollow noopener" data-state="closed"><a rel="noopener" class="inline" target="_blank" href="https://temporal.io/blog/durable-distributed-asyncio-event-loop"><span class="relative -mt-px select-none whitespace-nowrap -top-px font-sans text-base text-foreground selection:bg-super/50 selection:text-foreground dark:selection:bg-super/10 dark:selection:text-super"><span class="text-3xs rounded-badge group min-w-4 cursor-pointer text-center align-middle font-mono tabular-nums py-[0.1875rem] leading-snug px-[0.3rem] [@media(hover:hover)]:hover:bg-super dark:[@media(hover:hover)]:hover:text-inverse [@media(hover:hover)]:hover:text-white border-subtlest ring-subtlest divide-subtlest bg-subtle"><span class="relative -mt-px inline-block align-middle max-w-[25ch] overflow-hidden">temporal</span><span class="ml-xs -mt-px mr-px inline-block align-middle"><span class="opacity-50">+3</span></span></span></span></a></span></p>
<p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">To summarize:</p>
<div class="group relative"><div class="w-full overflow-x-auto md:max-w-[90vw] border-subtlest ring-subtlest divide-subtlest bg-transparent">
Aspect | asyncio.sleep (normal Python) | workflow.sleep / asyncio.sleep (Temporal workflow)
-- | -- | --
Implementation | Non-blocking async delay using OS event loop | Durable Temporal timer integrated into workflow event handling
Thread blocking | No | No
Durability | No (process crash loses state) | Yes (workflow state is replayed and resumed after crash)
Resource usage | Uses event loop resources | Does not consume worker resources while sleeping
Function availability | asyncio.sleep | Use asyncio.sleep; no separate workflow.sleep in Temporal Python

</div><div class="bg-base border-subtler shadow-subtle pointer-coarse:opacity-100 right-xs absolute bottom-0 flex rounded-lg border opacity-0 transition-opacity group-hover:opacity-100 [&amp;&gt;*:not(:first-child)]:border-subtle [&amp;&gt;*:not(:first-child)]:border-l"><div class="flex" style="opacity: 1;"><button aria-label="Copy table" type="button" class="focus-visible:bg-subtle text-quiet hover:text-foreground aspect-square font-sans focus:outline-none outline-none outline-transparent transition duration-300 ease-out select-none items-center relative group/button font-semimedium justify-center text-center items-center rounded-lg cursor-pointer active:scale-[0.97] active:duration-150 active:ease-outExpo origin-center whitespace-nowrap inline-flex text-sm h-8 aspect-[9/8]" data-state="closed"><div class="flex items-center min-w-0 gap-two justify-center"><div class="flex shrink-0 items-center justify-center size-4"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="tabler-icon tabler-icon-copy tabler-icon"><path d="M7 7m0 2.667a2.667 2.667 0 0 1 2.667 -2.667h8.666a2.667 2.667 0 0 1 2.667 2.667v8.666a2.667 2.667 0 0 1 -2.667 2.667h-8.666a2.667 2.667 0 0 1 -2.667 -2.667z"></path><path d="M4.012 16.737a2.005 2.005 0 0 1 -1.012 -1.737v-10c0 -1.1 .9 -2 2 -2h10c.75 0 1.158 .385 1.5 1"></path></svg></div></div></button></div><div class="flex" style="opacity: 1;"><button aria-label="Download CSV" type="button" class="focus-visible:bg-subtle text-quiet hover:text-foreground aspect-square font-sans focus:outline-none outline-none outline-transparent transition duration-300 ease-out select-none items-center relative group/button font-semimedium justify-center text-center items-center rounded-lg cursor-pointer active:scale-[0.97] active:duration-150 active:ease-outExpo origin-center whitespace-nowrap inline-flex text-sm h-8 aspect-[9/8]" data-state="closed"><div class="flex items-center min-w-0 gap-two justify-center"><div class="flex shrink-0 items-center justify-center size-4"><svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="tabler-icon tabler-icon-download tabler-icon"><path d="M4 17v2a2 2 0 0 0 2 2h12a2 2 0 0 0 2 -2v-2"></path><path d="M7 11l5 5l5 -5"></path><path d="M12 4l0 12"></path></svg></div></div></button></div></div></div>
<p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:pb-2">In practice, for Temporal Python SDK workflows, just use await asyncio.sleep to achieve the durable sleep behavior that is consistent with workflow execution semantics.<span class="citation inline" rel="nofollow noopener" data-state="closed"><a rel="noopener" class="inline" target="_blank" href="https://github.com/temporalio/sdk-python"><span class="relative -mt-px select-none whitespace-nowrap -top-px font-sans text-base text-foreground selection:bg-super/50 selection:text-foreground dark:selection:bg-super/10 dark:selection:text-super"><span class="text-3xs rounded-badge group min-w-4 cursor-pointer text-center align-middle font-mono tabular-nums py-[0.1875rem] leading-snug px-[0.3rem] [@media(hover:hover)]:hover:bg-super dark:[@media(hover:hover)]:hover:text-inverse [@media(hover:hover)]:hover:text-white border-subtlest ring-subtlest divide-subtlest bg-subtle"><span class="relative -mt-px inline-block align-middle max-w-[25ch] overflow-hidden">github</span><span class="ml-xs -mt-px mr-px inline-block align-middle"><span class="opacity-50">+3</span></span></span></span></a></span></p></div>

<ol>
  <li><a href="https://temporal.io/blog/durable-distributed-asyncio-event-loop">https://temporal.io/blog/durable-distributed-asyncio-event-loop</a></li>
  <li><a href="https://community.temporal.io/t/are-each-workflow-sleep-and-worflow-await-backed-by-a-thread-if-so-when-will-the-thread-be-relinquished/1526">https://community.temporal.io/t/are-each-workflow-sleep-and-worflow-await-backed-by-a-thread-if-so-when-will-the-thread-be-relinquished/1526</a></li>
  <li><a href="https://github.com/temporalio/sdk-python">https://github.com/temporalio/sdk-python</a></li>
  <li><a href="https://community.temporal.io/t/if-a-workflow-sleep-does-it-also-sleep-spawned-activities/6221">https://community.temporal.io/t/if-a-workflow-sleep-does-it-also-sleep-spawned-activities/6221</a></li>
  <li><a href="https://community.temporal.io/t/how-to-use-sleep-in-python-sdk/8773">https://community.temporal.io/t/how-to-use-sleep-in-python-sdk/8773</a></li>
  <li><a href="https://community.temporal.io/t/workflow-sleep-details/1161">https://community.temporal.io/t/workflow-sleep-details/1161</a></li>
  <li><a href="https://docs.temporal.io/develop/python/timers">https://docs.temporal.io/develop/python/timers</a></li>
  <li><a href="https://javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/workflow/Workflow.html">https://javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/workflow/Workflow.html</a></li>
  <li><a href="https://community.temporal.io/t/canceling-updating-a-temporal-sleep/12506">https://community.temporal.io/t/canceling-updating-a-temporal-sleep/12506</a></li>
  <li><a href="https://stackoverflow.com/questions/56729764/asyncio-sleep-vs-time-sleep">https://stackoverflow.com/questions/56729764/asyncio-sleep-vs-time-sleep</a></li>
</ol></body></html>
